Proposal implementation

arepnikov opened the pull request:
# Reflect Library Improvements Proposal

## Background

In Eventide's libraries, Reflect has 6 use cases, but 4 of them do not utilize it in the designed way. Instead of making a use of the reflection's actuator, the constant is pulled from the reflection.

The actuator is designed to pass the subject into the method being invoked. The subject can also be overridden, but either way, exactly one argument is always supplied to the protocol method:

```ruby
class SomeThing
  module Validate
    def self.call(some_thing)
      # validate some_thing
    end
  end
end

some_thing = SomeThing.new

validator_reflection = Reflect.(some_thing, :Validate)

validator_reflection.(:call) # validates some_thing

some_other_thing = SomeThing.new
validator_reflection.(:call, some_other_thing) # validates some_other_thing
```

However, the actuator doesn't support methods with zero arity or methods with arity greater than one. Due to this limitation, the constant is often pulled from the reflection, bypassing the actuator.

If the reflection's actuator were to support invoking protocol methods with any arity, then most use cases of the reflect library would be able to use the actuator and avoid disencapsulating the reflection's target constant.

## Reflection's Actuator

### Example: Method Without Arguments

#### Current Approach (Workaround)

```ruby
class SomeClass
  module Substitute
    def self.build
      # build the substitute
    end
  end
end

some_instance = SomeClass.new
some_reflection = Reflect.(some_instance, :Substitute)

constant = some_reflection.constant
constant.build
# => some substitute
```

#### Proposed Approach

```ruby
class SomeClass
  module Substitute
    def self.build
      # build the substitute
    end
  end
end

some_instance = SomeClass.new
some_reflection = Reflect.(some_instance, :Substitute)

some_reflection.(:build)
# => some substitute
```

### Example: Method With More Than One Argument

#### Current Way Around

```ruby
class SomeThing
  module Validate
    def self.call(some_thing, state)
      # validate some_thing
    end
  end
end

some_thing = SomeThing.new
validator_reflection = Reflect.(some_thing, :Validate)

validator = validator_reflection.constant
validator.call(some_thing, :some_state)
# => validate some_thing
```

#### Proposed Change

```ruby
class SomeThing
  module Validate
    def self.call(some_thing, state)
      # validate some_thing
    end
  end
end

some_thing = SomeThing.new
validator_reflection = Reflect.(some_thing, :Validate)

validator_reflection.(:call, some_thing, :some_state)
# => validate some_thing
```

## Reflection's Variant Actuator

Support for methods with any parameter signature is proposed. This necessitates a way to invoke the reflection's actuator so that it supplies the reflection's subject into the protocol method.

To address this, the introduction of a variant actuator method `.!()` is proposed. The variant actuator will pass the subject into the invoked method as the first positional argument, but will otherwise behave the same as the primary actuator:

```ruby
class SomeThing
  module Validate
    def self.call(some_thing, state)
      # validate some_thing
    end
  end
end

some_thing = SomeThing.new
validator_reflection = Reflect.(some_thing, :Validate)

# validator_reflection.(:call, some_thing, :some_state)
validator_reflection.!(:call, :some_state)
# => validate some_thing
```

# Additional Changes

## Target Method Predicate

``` ruby
class SomeThing
  module Validate
    def self.call(some_thing, state=nil)
      # validate some_thing, taking an account optional state
    end
  end
end

some_thing = SomeThing.new
validate_reflection = Reflect.(some_thing, :Validate)

validate_reflection.target_method?(:call)
# => true
```

## Arity Method

``` ruby
class SomeThing
  module Validate
    def self.call(some_thing, state=nil)
      # validate some_thing, taking an account optional state
    end
  end
end

some_thing = SomeThing.new
validate_reflection = Reflect.(some_thing, :Validate)

validate_reflection.arity(:call)
# => -2
```

## Parameters Method

``` ruby
class SomeThing
  module Validate
    def self.call(some_thing, state=nil)
      # validate some_thing, taking an account optional state
    end
  end
end

some_thing = SomeThing.new
validate_reflection = Reflect.(some_thing, :Validate)

validate_reflection.parameters(:call)
# => [[:some_thing, :req], [:state, :opt]]
```

arepnikov committed:
Proposal of Reflect improvements in markdown doc

arepnikov committed:
Proposal of Reflect more examples and improvements

arepnikov committed:
Actuate stale tests removed

arepnikov committed:
RecordInvocation development dependency added

arepnikov committed:
Improvements for Reflect library - tests are added

arepnikov committed:
Proposal implementation

arepnikov committed:
Missing tests added to arity and parameters methods

arepnikov committed:
Actuator and variant actuator tests clarified

ntl commented on lib/reflect/reflection.rb:
```diff
@@ -27,21 +27,42 @@ def self.build(subject, constant_name, strict: nil, ancestors: nil)
       subject ||= constant
       subject.respond_to?(name)
     end
+    alias :target_method? :target_accessor?
```
target_accessor should be implemented in terms of target_method, not the other way around

ntl replied:
Also, consider `def target_accessor?` rather than `alias`, since it's a distinct member of Reflection's protocol

ntl commented on lib/reflect/reflection.rb:
```diff
@@ -27,21 +27,42 @@ def self.build(subject, constant_name, strict: nil, ancestors: nil)
+      end
+
+      target.public_method(method_name).parameters
+    end
```
Consider: either `assure_target_method(method_name)` or `target_method(method_name)`

ntl commented on proposal.md:
```diff
@@ -0,0 +1,283 @@
+
+In Eventide's libraries, Reflect has 6 use cases, but 4 of them do not utilize it in the designed way. Instead of making a use of the reflection's actuator, the constant is pulled from the reflection.
+
+The actuator is designed to pass the subject into the method being invoked. Subject could be overriden when needed.
```
Consider for the second sentence: "The subject can also be overridden, but either way, exactly one argument is always supplied to the protocol method:"

(The colon is to indicate that the following code pertains to the paragraph)

ntl commented on proposal.md:
````diff
@@ -0,0 +1,283 @@
+validator_reflection = Reflect.(some_thing, :Validate)
+
+validator_reflection.(:call)
+```
````
Consider demonstrating the argument override as well

ntl commented on proposal.md:
````diff
@@ -0,0 +1,283 @@
+validator_reflection.(:call)
+```
+
+However, the actuator doesn't support methods with zero arity or methods with arity greater than one. Due to these limitations, the constant is extracted from the reflection, bypassing the actuator.
````
Consider: "Due to this limitation, the constant is often pulled from the reflection, ..."

ntl commented on proposal.md:
```diff
@@ -0,0 +1,283 @@
+
+However, the actuator doesn't support methods with zero arity or methods with arity greater than one. Due to these limitations, the constant is extracted from the reflection, bypassing the actuator.
+
+It has been observed that the reflection's actuator should support methods with any arity and allow the subject to be overridden when needed.
```
Consider: If the reflection's actuator were to support invoking protocol methods with any arity, then most use cases of the reflect library would be able to use the actuator and avoid disencapsulating the reflection's target constant.

ntl commented on proposal.md:
```diff
@@ -0,0 +1,283 @@
+
+It has been observed that the reflection's actuator should support methods with any arity and allow the subject to be overridden when needed.
+
+## Problem - Reflection's Actuator Only Supports Methods With Precisly One Argument
```
Consider: "Example" rather than "Problem"

ntl commented on proposal.md:
```diff
@@ -0,0 +1,283 @@
+      puts 'some substitute'
+    end
+  end
+end
```
Eliminate the redundant code

ntl commented on proposal.md:
````diff
@@ -0,0 +1,283 @@
+```
+
+
+### Method With More Than One Argument
````
Consider adding Example as above

ntl commented on proposal.md:
```diff
@@ -0,0 +1,283 @@
+
+## Reflection's Variant Actuator - Passing Subject and Methods With Any Parameter Signature
+
+Support for methods with any parameter signature is proposed. This necessitates a way to instruct the reflection on whether it should pass the subject into the invoked method.
```
Consider: "This necessitates a way to invoke the reflection's actuator so that it supplies the reflection's subject into the protocol method."

ntl commented on proposal.md:
```diff
@@ -0,0 +1,283 @@
+
+Support for methods with any parameter signature is proposed. This necessitates a way to instruct the reflection on whether it should pass the subject into the invoked method.
+
+To address this, the introduction of a variant actuator method `.!()` is proposed. The variant actuator invariably will pass the subject into the invoked method.
```
Consider: "The variant actuator will pass the subject into the invoked method as the first positional argument, but will otherwise behave the same as the primary actuator."

ntl commented on proposal.md:
````diff
@@ -0,0 +1,283 @@
+validator_reflection.!(:call, :some_state)
+# => validate some_thing
+```
+
````
Consider replacing the rest of this documentation with a section titled, "Additional Changes," that covers the additional methods introduced by this pull request (e.g. `target_method?` and `target_method`)

ntl commented on test/automated/reflection/actuator/actuator.rb:
```diff
@@ -0,0 +1,47 @@
+  context "Reflection" do
+    context "Actuator" do
+      subject = Reflect::Controls::Subject::Example.new
+      target = Reflect::Controls::Subject::Example::SomeConstant
```
Consider:

``` ruby
constant_name = Controls::Namespace::Random.get
subject = Controls::Subject::Example.new
target = Reflect::Controls::Target.example(constant_name:, subject:)
```

Possible implementation:

``` ruby
module Reflect
  module Controls
    module Target
      def self.example(constant_name: nil, subject: nil)
        constant_name ||= Namespace::Random.get
        subject ||= Subject.example

        target = Module.new do
          set_temporary_name(constant_name)
          
          extend Methods
        end
        
        subject.class.const_set(constant_name, target)
        
        target
      end
      
      module Methods
        # Somehow has the same methods as Subject::Example::SomeConstant
      end
    end
  end
end
```

ntl commented on test/automated/reflection/actuator/actuator.rb:
```diff
@@ -0,0 +1,47 @@
+      reflection.(method_name, subject, :some_arg)
+
+      invocations = target.invocations(method_name)
+      invocation = invocations.last
```
Getting the first invocation should be possible after the above changes

ntl commented on test/automated/reflection/actuator/mixed_parameters.rb:
```diff
@@ -0,0 +1,105 @@
+    context "Actuator" do
+      context "Mixed Parameters" do
+        subject = Reflect::Controls::Subject::MixedParameters::Example.new
+        target  = Reflect::Controls::Subject::MixedParameters::Example::SomeConstant
```
Consider:

``` ruby
target = Controls::Target::MixedParameters.example(constant_name:, subject:)
```

ntl commented on test/automated/reflection/actuator/mixed_parameters.rb:
```diff
@@ -0,0 +1,105 @@
+        )
+
+        invocations = target.invocations(method_name)
+        invocation = invocations.last
```
After above changes, getting first invocation should be possible

ntl commented on test/automated/reflection/arity/method_is_not_implemented.rb:
```diff
@@ -0,0 +1,21 @@
+      end
+    end
+  end
+end
```
Consider: after `target_method` is implemented with test coverage for when the method isn't implemented, remove this file since it's test coverage is redundant. Expect to have just a single `arity.rb` test file with no `arity` directory afterwards.

ntl commented on test/automated/reflection/parameters/method_is_not_implemented.rb:
```diff
@@ -0,0 +1,21 @@
+      end
+    end
+  end
+end
```
Consider: after target_method is implemented with test coverage for when the method isn't implemented, remove this file since it's test coverage is redundant. Expect to have just a single `parameters.rb` test file with no `parameters` directory afterwards.

ntl commented on test/automated/reflection/parameters/method_is_not_implemented.rb:
```diff
@@ -0,0 +1,21 @@
+
+        reflection = Reflect.(subject, :SomeConstant, strict: true)
+
+        random_method_name = Reflect::Controls::Namespace::Random.get
```
Consider `method_name` rather than `random_method_name`

ntl commented on test/automated/reflection/target_method_predicate.rb:
```diff
@@ -0,0 +1,31 @@
+    context "Target Method Predicate" do
+      context "Method Implemented" do
+        subject = Reflect::Controls::Subject.example
+        reflection = Reflect.(subject, :SomeConstant, strict: true)
```
Consider: don't supply `strict: true` since it has no bearing on the test, and should be tested elsewhere (verify this!)

ntl replied:
Consider: `subject` and `reflection` can be assigned outside this context, and shared between the `Method Implemented` and `Method Is Not Implemented` context

ntl commented on test/automated/reflection/target_method_predicate.rb:
```diff
@@ -0,0 +1,31 @@
+        reflection = Reflect.(subject, :SomeConstant, strict: true)
+
+        random_method_name = Reflect::Controls::Namespace::Random.get
+        method_implemented = reflection.target_method?(random_method_name)
```
This is incorrect. A random namespace isn't the same thing as a random method name.

Consider: `method_name = :not_implemented_method`

ntl commented on test/automated/reflection/variant_actuator/method_is_not_implemented.rb:
```diff
@@ -0,0 +1,21 @@
+
+        reflection = Reflect.(subject, :SomeConstant, strict: true)
+
+        random_method_name = Reflect::Controls::Namespace::Random.get
```
Same as above re. `:not_implemented_method`.

Also, consider `Controls::MethodName.random` and `Controls::MethodName.example`. But `random_method_name` would not be correct in that case, because just `method_name` would suffice.

arepnikov committed:
target_method? rather than target_accessor?

target_accessor should be implemented in terms of target_method,
not the other way around.

def target_accessor? rather than alias, since it's a distinct member of
Reflection's protocol

arepnikov committed:
get_target with `public_send` rather than `send`

we can replace `send` with `public_send`, since before was called
`respond_to?` (without any optional arguments), which means that
respond_to? returned true only for existing public methods.

arepnikov committed:
assure_target_method added

arepnikov committed:
proposal clarified

arepnikov committed:
Target controls added and used in Subject example

arepnikov committed:
MixedParameters test clarified

arepnikov committed:
Actuator and Variant Actuator test clarified

arepnikov committed:
Parameters test clarified

arepnikov committed:
target_method introduced

arepnikov committed:
redundant method_is_not_implemented tests removed

arepnikov committed:
arity and parameters tests moved

arepnikov committed:
redundant strict argument removed

arepnikov committed:
arity and target_method? tests clarified

ntl commented on lib/reflect/controls/target.rb:
```diff
@@ -0,0 +1,95 @@
+
+          def some_accessor
+            some_inner_module = const_get(:SomeInnerConstant)
+            some_inner_module
```
The difference between `SomeInnerConstant` and `some_inner_module` needs an explanation (or to be harmonized if there isn't a need for a difference)

ntl commented on lib/reflect/controls/target.rb:
```diff
@@ -0,0 +1,95 @@
+          def self.extended(cls)
+            cls.extend(SomeMethod)
+            cls.extend(Target::Methods::SomeAccessor)
+            cls.extend(Target::Methods::SomeObjectAccessor)
```
If these aren't used, they must not be included

ntl commented on lib/reflect/reflection.rb:
```diff
@@ -27,20 +27,41 @@ def self.build(subject, constant_name, strict: nil, ancestors: nil)
+    end
+
+    def !(method_name, ...)
+      call(method_name, subject, ...)
```
Consider: `self.(` rather than `call(`

ntl commented on lib/reflect/reflection.rb:
```diff
@@ -27,20 +27,41 @@ def self.build(subject, constant_name, strict: nil, ancestors: nil)
+    def target_method?(method_name, subject=nil)
+      subject ||= target
+      subject.respond_to?(method_name)
+    end
```
Remove optional subject argument and related tests

ntl commented on lib/reflect/reflection.rb:
```diff
@@ -27,20 +27,41 @@ def self.build(subject, constant_name, strict: nil, ancestors: nil)
+
+    def target_accessor?(method_name, subject=nil)
+      target_method?(method_name, subject)
+    end
```
Introduce comment:

```
## Review purpose of optional subject and consider its removal - Alek, Nathan, Mon Sep 16 2024
```

ntl commented on test/automated/reflection/actuator/actuator.rb:
```diff
@@ -0,0 +1,45 @@
+
+            test "Value" do
+              assert(value == subject)
+            end
```
Run the test and observe whether the test needs a title at all (currently it's `"Value"`)

ntl commented on test/automated/reflection/actuator/actuator.rb:
```diff
@@ -0,0 +1,45 @@
+      RecordInvocation.(target, method_name)
+
+      reflection = Reflect.(subject, constant_name)
+      reflection.(method_name, subject, :some_arg)
```
Consider: `reflect` as the variable name. Rationale: `some_noun.()` doesn't make sense but `some_verb.()` does. (`some_noun.some_verb` also makes sense but isn't relevant here)

ntl commented on test/automated/reflection/actuator/mixed_parameters.rb:
```diff
@@ -0,0 +1,103 @@
+        method_name = :some_method
+        RecordInvocation.(target, method_name)
+
+        control_block = proc { nil }
```
Consider: `proc { }` rather than `proc { nil }`

arepnikov committed:
Target control clarified

arepnikov committed:
target_accessor? method reverted to master state and comment added

arepnikov committed:
target_method? clarified by removing redundant optional argument

arepnikov committed:
porposam removed and contnet used as pull request desciption
